### PR TITLE
Fail if field lengths are not same in INTERSECT and EXPECT

### LINKED ISSUE #3632: `INTERSECT` and `EXCEPT` don't return an error when 2 sets have the different number of columns
**Describe the bug**

**To Reproduce**
```sql
❯ select 1 as a, 2 as b
intersect
select 1 as a;
+---+---+
| a | b |
+---+---+
| 1 | 2 |
+---+---+
1 row in set. Query took 0.009 seconds.
❯ select 1 as a, 2 as b
except
select 2 as a;
+---+---+
| a | b |
+---+---+
| 1 | 2 |
+---+---+
```

**Expected behavior**
We should return an error, which is the behaviour in the Postgresql: https://onecompiler.com/postgresql/3yhau7n8a

**Additional context**

### PR DESCRIPTION
**Pull request by askoa**
# Which issue does this PR close?

Closes #3632

# Are there any user-facing changes?

Users will see failure if they are using INTERSECT and EXPECT on tables with different field count.

### CODE REVIEW

File: datafusion/expr/src/logical_plan/builder.rs [hunk 1]
```diff
@@ -781,6 +781,16 @@ impl LogicalPlanBuilder {
         join_type: JoinType,
         is_all: bool,
     ) -> Result<LogicalPlan> {
+        let left_len = left_plan.schema().fields().len();
+        let right_len = right_plan.schema().fields().len();
+
+        if left_len != right_len {
+            return Err(DataFusionError::Plan(format!(
+                "Expected the no. of fields to be the same. Left fields count = {} and right fields count = {} are not the same.",
```
Comment (xudong963) [inline datafusion/expr/src/logical_plan/builder.rs:789]:
```suggestion
                "each INTERSECT/EXCEPT query must have the same number of columns. Left fields count = {} and right fields count = {} are not the same.",
```

File: datafusion/core/tests/sql/intersection.rs [hunk 2]
```diff
@@ -85,3 +85,19 @@ async fn test_intersect_distinct() -> Result<()> {
     assert_batches_eq!(expected, &actual);
     Ok(())
 }
+
+#[tokio::test]
+#[should_panic(
```
Comment (xudong963) [inline datafusion/core/tests/sql/intersection.rs:90]: You can reference the test in https://github.com/apache/arrow-datafusion/pull/3638/files which is more elegant.

File: datafusion/expr/src/logical_plan/builder.rs (hunk 1 above)
Comment (askoa) [reply datafusion/expr/src/logical_plan/builder.rs]: I updated based on your suggestion. Let me know if that works.

File: datafusion/core/tests/sql/intersection.rs (hunk 2 above)
Comment (askoa) [reply datafusion/core/tests/sql/intersection.rs]: done!

**Review by xudong963 — APPROVED** [review]
Thanks @askoa

**Comment by askoa** [comment]
CI failure unrelated to this PR.

**Review by HaoYang670 — APPROVED** [review]
LGTM. Thank you, @askoa !

**Pull request merged**